Added data classes to Zone

manzanotti opened the pull request:
Added tests for all these data classes.

Also includes an attempt to get the tests to run on CircleCI.

manzanotti committed:
Added data classes to Zone

Added tests for all these data classes.

Also includes an attempt to get the tests to run on CircleCI.

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -33,15 +34,128 @@ def alphanum_key(k):
+        self.update_warmup(json)
+        self.update_override(json)
+        self.update_trigger(json)
+        self.update_is_occupied(json)
```
Not all zones support all schedules (if at all):
https://github.com/zxdavb/geniushub-client/blob/b419463f6eba2436753777025faab8d7cf83f928/geniushubclient/zone.py#L149-L163

zxdavb replied:
If you look in my `data()`, you'll see the above `if` statements. 

BTW, ignore my dodgy `return`

manzanotti replied:
Added this missing logic in. Not sure how I missed it originally!

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -221,18 +344,257 @@ def footprint_schedule(raw_json) -> Dict:
 
         return self._data
 
+    def update_state(self, json):
```
Maybe these `def update_...()` should be: `def _update_...()`?

manzanotti replied:
The main reason that I left them was that I wasn't sure if these methods would eventually be able to be called individually by a client, hence needing them to not be private. However, there's no reason that they couldn't be (as they're not really private), and if a use case ever comes up for them needing to be public, it can be changed then.

manzanotti replied:
Changed these all to private methods.

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -221,18 +344,257 @@ def footprint_schedule(raw_json) -> Dict:
+    def update_timer_schedule(self, json):
+        """Parse the json and use it to populate the timer_schedule data class."""
+        try:
+            dayIndex = -1
```
snake_case pls, say: `day_idx`

manzanotti replied:
Sorry, camel case is something I'm so used to doing, it's what C#, Javascript and TypeScript all se. Updated!

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -33,15 +34,128 @@ def alphanum_key(k):
+class DaySchedule:
+    """The day in a schedule. Contains a list of heating periods."""
+
+    index: int
```
Should have initial values for all of these? `index: int = 0` Or does it set it to `None`? (I guess I could look that up myself)

manzanotti replied:
Done! Also made the name a property that is calculated from the index, rather than making the calling code do the calculation. Initialised the List too, so now the constructor only requires two parameters sending in.

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -33,15 +34,128 @@ def alphanum_key(k):
+    """
+
+    type: int = 0
+    type_name: str = ""
```
Why have both `type`, and `type_name` - one is F(x) of the other. See also below: `day_idx`, and `day_name`?

And we should be careful about using type as an attribute name (type() is 'reserved') - maybe `zone_type`.

manzanotti replied:
Oh, yes, should have updated this too. I've updated the day name field to be a property that does the conversion itself. Will do that here too.

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -221,18 +344,257 @@ def footprint_schedule(raw_json) -> Dict:
+            self._properties.has_room_sensor = json["iFlagExpectedKit"] & ZONE_KIT.PIR
+
+        except (AttributeError, LookupError, TypeError, ValueError):
+            _LOGGER.exception("Failed to set Zone %s properties.", self.id)
```
https://github.com/zxdavb/geniushub-client/blob/b419463f6eba2436753777025faab8d7cf83f928/geniushubclient/zone.py#L389-L398

zxdavb replied:
I couldn't help myself - I refactored much of these (I like very precise code)- please don't take offense!

manzanotti replied:
No offense taken! Do have some questions about some of these changes though. I'll ask them in gitter.

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -221,18 +344,257 @@ def footprint_schedule(raw_json) -> Dict:
+                self._state.is_occupied = False
+                return
+
+            self._state.is_occupied = self._footprint.reactive.activity_level > 0.0
```
https://github.com/zxdavb/geniushub-client/blob/b419463f6eba2436753777025faab8d7cf83f928/geniushubclient/zone.py#L423-L434

zxdavb replied:
Another example.

manzanotti replied:
Ah, did not know about elif, that's much neater!

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -221,18 +344,257 @@ def footprint_schedule(raw_json) -> Dict:
+
+            if iType == ZONE_TYPE.Manager:
+                if json["fPV"]:
+                    self._state.temperature = json["fPV"]
```
Have we got a problem here? In the v1 API JSON, you get something like this if zone 0 does not have a thermostat:
`{
  "name": "Whole House",
  ...
}`

Or, if it does have a thermostat, t would be:
`{
  "name": "Whole House",
  "temperature": 17.8,
  ...
}`

Will this code instead do:
`{
  "name": "Whole House",
  "temperature": null,
  ...
}`
Because there should be no `"temperature"` key at all!

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -221,18 +344,257 @@ def footprint_schedule(raw_json) -> Dict:
+        """Parse the json and use it to populate the properties data class."""
+        try:
+            iType = json["iType"]
+            self._properties.type = iType
```
In my little refactor, I may have changed this to:
```
self._properties.type = iType = json["iType"]
```

manzanotti replied:
Done.

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -33,15 +34,128 @@ def alphanum_key(k):
+    mode: int = 0
+    mode_name: str = ""
+    temperature: float = 0.0
+    set_point: float = 0.0
```
`setpoint`

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -33,15 +34,128 @@ def alphanum_key(k):
+
+    type: int = 0
+    type_name: str = ""
+    has_room_sensor: bool = False
```
We must be careful - since this data don't exist in the v3 API, we're establishing a namespace here that may not last the next revision of by the vendor - that's why I used `_has_pir`.

manzanotti replied:
Ah, missed this one in my update. I must admit, I assumed that has_pir basically meant whether the room has a sensor, so I just wanted to make that more obvious. Does it have other connotations then?

zxdavb replied:
There are no connotations to `has_pir`.

It's the leading _ which is the main thing (changing the name is only a 2' issue) - I was using the leading _ to imply a name private to the library - use it if you will, but it's not a public name, and no promises made it will hang around (since it is not in the public API yet, and so may change).

My preference is to keep the name for now only for the usual reasons (HA code currently uses it, it may have to change again in the future), and the leading _ for the reasons above.

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -221,18 +344,257 @@ def footprint_schedule(raw_json) -> Dict:
+
+    def update_properties(self, json):
+        """Parse the json and use it to populate the properties data class."""
+        try:
```
Ideally we should significantly minimize use of try/except
 - least code inside t/e block as possible - move some outside where we can  
 - maybe even only wrap the complex modules - timer & schedule?
 - least number of exceptions as possible (having said that, do we need to add NameError now)
 - can be used for forgiveness as well as error handling

manzanotti replied:
It's a fair comment on only having these on the complex functions. I saw that those original code blocks had them in, so added them to all the update methods to be consistent. Also, as this is all new, I thought that it was better for me to err on the side of caution with all these.

I'm happy to reduce the number, if you think that's better. Would you want one over-arching one in the update method, as well as the two complex functions?

zxdavb replied:
I don't profess to be a Python expert (and I know nothing about java, et al.), but I wonder if there is a difference in philosophy about the these try/catch, try/except clauses? Anyway...

Let's be brave - wrap complex code (timer, footprint), but no over-arching try/except. We can do extensive testing as is essentially Summer now - anyone who is risk-averse can stick to the PyPi library for now.

manzanotti replied:
No, in C# you really shouldn't use too many try/catch blocks, as exceptions are intensive to create (the whole stack needs to be unwound to create the stack trace, for starters). As I said, I only added them so that all the update methods were consistent, but they were a pain in the arse for the unit tests, as they would catch the errors, so tests could pass when they shouldn't.

I'll just keep the try/catch code in the schedule update methods.

manzanotti replied:
Updated to remove the extra error catching.

zxdavb commented on geniushubclient/zone.py:
```diff
@@ -33,15 +34,128 @@ def alphanum_key(k):
+        self.update_is_occupied(json)
+
     @property
     def data(self) -> Dict:
```
I'm not sure why this has been left in - all the main JSON conversion code is now duplicated? We should be able to do something similar to:
https://github.com/zxdavb/geniushub-client/blob/b419463f6eba2436753777025faab8d7cf83f928/geniushubclient/zone.py#L167-L176

zxdavb replied:
This may be easier if we are careful with the dataclass' namespaces.

manzanotti replied:
I presume that you are referring to the data property? I left it in so that my changes were purely additive, I figured that as it's a very large change, it would be a lot safer to get the new stuff in, be happy that it works, then remove the old conversion code.

Happy to delete it if you are!

zxdavb replied:
Please take it out.

manzanotti replied:
See my reply in Gitter as to why I left this in.

zxdavb commented:
@manzanotti I am so sorry to have taken so long to review your very useful PR (many reasons).

I can see where you're heading & it looks great!

zxdavb reviewed (changes requested):
Main thing is that I have added a [branch](https://github.com/zxdavb/geniushub-client/tree/pr-38), which is based upon your #38 - the reason why I did this is because (blah-blah, and) I have no idea how to send a commit direct to you... Oh well.

Maybe you might rebase off that?

The code didn't run on my system without error, the main reason being that not all zone types support all schedule types - see my review notes. 

And sorry, I did a bit of refactoring - I couldn't stop myself!!

manzanotti committed:
Corrections following code review

All data class properties now have default values. DaySchedule construction tidied up.
Corrected camel case variable names.
Added missing business logic for json processing.
update_is_occupied now uses elif blocks rather than separate if statements with returns.

manzanotti committed:
Removed error catching from update methods

Also made update methods private

manzanotti committed:
Added tests for the data method.
